### PR TITLE
Add Worker Skeleton, BaseSingerWorker, and SingerDiscoveryWorker

### PR DESCRIPTION
**Pull request by sherifnada**
This is a draft of the approach we can take to implement workers to get early feedback. Definitely not a complete PR. 

TL;DR This PR adds an initial framework for workers and singer workers.

Read files in this order: 

- `IWorker.java`: This is the basic contract of a worker. 
- `WorkerStatus`: part of the contract of a worker
- `BaseSingerWorker`: Abstract class containing common functionality to all singer workers. The intention of this class is to handle all workers which are 100% singer (singer test, singer discovery, and singer-to-singer-sync). Once we move to Singer/non-singer sync, this class might need to change or more likely be wrapped in something else. 
- `SingerDiscoveryWorker`: an example of how a singer worker would use the abstract class. 
- everything else in no particular order

This contains some basic tests in `TestBaseSingerWorker` which verify that states are transitioned correctly and that singer workers can be cancelled. More worker-specific tests will come once #44 is resolved.

TODO:

- [ ] failing test

### CODE REVIEW

**Review by cgardens — APPROVED** [review]
a bunch of questions and thoughts. please take a look at them and share your answers, but don't block merging waiting for me to respond to your answers to the questions. i think this directionally the right thing and i most just want to hear you articulate more how you're thinking about a couple things.

where possible if you can add clarification on the ones that are asking about "how stuff works" that would be great!

nice work.

**Review by jrhizor — CHANGES_REQUESTED** [review]
changing mine to request changes due to Charles

**Review by jrhizor — APPROVED** [review]
